File: backend/src/gitService.js
import { spawnSync } from "child_process";
import { existsSync } from "fs";
import { join } from "path";
import { settings } from "./config.js";

export class GitError extends Error {}

export class GitService {
  constructor(logs, pat = null) {
    this.logs = logs;
    this.pat = pat;
    this.repoPath = settings.repoMountPath;
  }

  env() {
    const env = { ...process.env };
    if (this.pat) {
      env.GIT_TERMINAL_PROMPT = "0";
      env.GIT_ASKPASS = "echo";
      env.GIT_USERNAME = "pat";
      env.GIT_PASSWORD = this.pat;
    }
    return env;
  }

  run(args, logCmd = true) {
    if (logCmd) this.logs.info(`git ${args.join(" ")}`);
    const result = spawnSync("git", args, {
      cwd: this.repoPath,
      encoding: "utf-8",
      env: this.env(),
    });
    if (result.status !== 0) {
      const message =
        (result.stderr || result.stdout || "").trim() || "Commande git échouée";
      this.logs.error(message);
      throw new GitError(message);
    }
    const out = (result.stdout || "").trim();
    if (out && logCmd) {
      const last = out.split("\n").pop();
      this.logs.info(last.slice(0, 200));
    }
    return { stdout: result.stdout || "", stderr: result.stderr || "" };
  }

  validateRepo() {
    if (!existsSync(this.repoPath)) {
      throw new GitError(`Repo introuvable: ${this.repoPath}`);
    }
    if (!existsSync(join(this.repoPath, ".git"))) {
      throw new GitError(`Pas un dépôt git: ${this.repoPath}`);
    }
  }

  lsRemote() {
    this.run(["ls-remote", settings.remoteName, "HEAD"]);
  }

  checkout(branch) {
    this.run(["checkout", branch]);
  }

  fetch() {
    this.run(["fetch", settings.remoteName]);
  }

  pull(branch = settings.defaultBranch) {
    this.run(["pull", settings.remoteName, branch]);
  }

  createBranch(name) {
    this.run(["checkout", "-b", name]);
  }

  commit(message) {
    this.run(["add", "-A", settings.targetFolder]);
    const status = this.run(["status", "--porcelain"], false);
    if (!status.stdout.trim()) {
      throw new GitError("Aucun changement à committer");
    }
    this.run(["commit", "-m", message]);
    return this.run(["rev-parse", "--short", "HEAD"], false).stdout.trim();
  }

  push(branchName) {
    this.run(["push", "-u", settings.remoteName, branchName]);
  }

  prepareBaseBranch() {
    this.checkout(settings.defaultBranch);
    this.fetch();
    this.pull(settings.defaultBranch);
    const head = this.run(["rev-parse", "--short", "HEAD"], false).stdout.trim();
    this.logs.info(`Branche ${settings.defaultBranch} à jour (${head})`);
    return head;
  }
}

File: backend/app/git_service.py
from __future__ import annotations

import os
import subprocess
from dataclasses import dataclass
from pathlib import Path
from typing import Callable

from app.config import Settings
from app.logging_service import LogLevel, SessionLogStore


class GitError(Exception):
    pass


@dataclass
class GitResult:
    stdout: str
    stderr: str


class GitService:
    def __init__(
        self,
        settings: Settings,
        logs: SessionLogStore,
        pat: str | None = None,
    ) -> None:
        self.settings = settings
        self.logs = logs
        self.pat = pat
        self.repo_path = Path(settings.repo_mount_path)

    def _env(self) -> dict[str, str]:
        env = os.environ.copy()
        if self.pat:
            env["GIT_TERMINAL_PROMPT"] = "0"
            env["GIT_ASKPASS"] = "echo"
            env["GIT_USERNAME"] = "pat"
            env["GIT_PASSWORD"] = self.pat
        return env

    def _run(self, args: list[str], *, log_cmd: bool = True) -> GitResult:
        if log_cmd:
            self.logs.info(f"git {' '.join(args)}")
        result = subprocess.run(
            ["git", *args],
            cwd=self.repo_path,
            capture_output=True,
            text=True,
            env=self._env(),
            check=False,
        )
        if result.returncode != 0:
            message = result.stderr.strip() or result.stdout.strip() or "Commande git échouée"
            self.logs.error(message)
            raise GitError(message)
        output = result.stdout.strip()
        if output:
            self.logs.info(output.splitlines()[-1][:200])
        return GitResult(stdout=result.stdout, stderr=result.stderr)

    def validate_repo(self) -> None:
        if not self.repo_path.exists():
            raise GitError(f"Repo introuvable: {self.repo_path}")
        if not (self.repo_path / ".git").exists():
            raise GitError(f"Pas un dépôt git: {self.repo_path}")

    def ls_remote(self) -> None:
        self._run(["ls-remote", self.settings.remote_name, "HEAD"])

    def current_branch(self) -> str:
        result = self._run(["branch", "--show-current"], log_cmd=False)
        return result.stdout.strip() or "HEAD"

    def checkout(self, branch: str) -> None:
        self._run(["checkout", branch])

    def fetch(self) -> None:
        self._run(["fetch", self.settings.remote_name])

    def pull(self, branch: str | None = None) -> None:
        branch = branch or self.settings.default_branch
        self._run(["pull", self.settings.remote_name, branch])

    def create_branch(self, branch_name: str) -> None:
        self._run(["checkout", "-b", branch_name])

    def commit(self, message: str) -> str:
        self._run(["add", "-A", self.settings.target_folder])
        status = self._run(["status", "--porcelain"], log_cmd=False)
        if not status.stdout.strip():
            raise GitError("Aucun changement à committer")
        self._run(["commit", "-m", message])
        sha = self._run(["rev-parse", "--short", "HEAD"], log_cmd=False).stdout.strip()
        return sha

    def push(self, branch_name: str) -> None:
        self._run(["push", "-u", self.settings.remote_name, branch_name])

    def prepare_base_branch(self) -> str:
        branch = self.settings.default_branch
        self.checkout(branch)
        self.fetch()
        self.pull(branch)
        head = self._run(["rev-parse", "--short", "HEAD"], log_cmd=False).stdout.strip()
        self.logs.info(f"Branche {branch} à jour ({head})")
        return head
